### PR TITLE
Revise the implementation of set_activation_checkpointing

### PR DESCRIPTION
**Pull request by rohan-varma**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [x] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

`set_activation_checkpointing` is a bit restrictive in that it only allows policies to be taken in via a defined `set` instead of allowing users to also author policies. We'll soon be improving our activation checkpointing to make llama3 training more memory efficient, which will require a more custom policy this PR will enable. More context in https://github.com/pytorch/torchtune/issues/781. Also improving some of the documentation around the API and removing allowing passing of `None` into the policy since this does not make sense

#### Changelog
- Improve documentation
- Stop allowing `None` passed as the policy
- Allow custom policy to be passed in, in addition to sets
- Add unittests 

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

`pytest test_memory.py -v`

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: torchtune/utils/memory.py [hunk 1]
```diff
@@ -20,19 +20,27 @@
...
-    """Utility to setup activation checkpointing and wrap the model for checkpointing.
+    """Utility to apply activation checkpointing to the passed in model.
 
     Args:
         model (nn.Module): Model to setup activation checkpointing.
-        auto_wrap_policy (Optional[Set[nn.Module]]): Policy to wrap module.
-        **kwargs: additional arguments to pass to torch.distributed activation checkpointing.
+        auto_wrap_policy (ACWrapPolicyType): Policy to wrap module.
+            This can either be a set of ``nn.Module`` types, in which case, modules of the specified type(s)
+            will be wrapped individually with activation checkpointing or a ``callable`` policy describing
```
Comment (ebsmothers) [inline torchtune/utils/memory.py:35]: nit: commas
```suggestion
            This can either be a set of ``nn.Module`` types, in which case modules of the specified type(s)
            will be wrapped individually with activation checkpointing, or a ``callable`` policy describing
```

File: tests/torchtune/utils/test_memory.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
+    CheckpointWrapper,
+)
+from torchtune.utils import set_activation_checkpointing
+
+
+class TestSetActivationCheckpointing:
+    @pytest.fixture
+    def model(self) -> int:
+        return nn.Sequential(nn.Linear(10, 10), nn.Linear(10, 10), nn.Linear(10, 10))
```
Comment (ebsmothers) [inline tests/torchtune/utils/test_memory.py:18]: nit: maybe something slightly more nontrivial here (i.e. not all nn.Linears)

**Review by ebsmothers — APPROVED** [review]
Couple small nits but looks good!
